### PR TITLE
Add types for the configuration object

### PR DESCRIPTION
**Pull request by matissepage**
### Why are we doing this feature?

Write the Typescript type of the Kuzzle config

### How did we make this feature?

avoid config errors, auto-completion, auto-documentation

### Explain: 

create some sub-type like a plugins type for exemple

### How should this be manually tested?

 ```docker-compose up```

### CODE REVIEW

File: lib/types/config/KuzzleConfiguration.ts [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+       * @default 0
+       */
+      accessLogIpOffset: number
+    },
+
+    protocols: Record<string, unknown>,
+  },
+}
+
+export type KuzzleConfiguration = Partial<IKuzzleConfiguration>
```
Comment (Aschen) [inline lib/types/config/KuzzleConfiguration.ts:633]: I didn't knew that, very nice :+1:

**Review by Aschen — CHANGES_REQUESTED** [review]
Files in the `types/config` directory should only be related to Kuzzle configuration. Even if you are using types like `RoleDefinition` in the configuration, those types usage are not limited to the configuration file and thus shouldn't be in the directory.

You could organize them in another way, feel free to propose something on this PR discussion before making changes if you want to discuss it beforehand

**Comment by Aschen** [comment]
You pull request markdown is broken

**Review by xbill82 — APPROVED** [review]

**Review by Aschen — APPROVED** [review]

**Review by Aschen — CHANGES_REQUESTED** [review]
Last remarks:

- `elasticsearch/StorageService.ts` it should rather be `storageService/StorageServiceElasticsearchConfiguration.ts`
- `InternalCacheConfiguration.ts` it should rather be `internalCache/InternalCacheRedisConfiguration.ts`
- `MemoryStorageConfiguration` it should rather be `publicCache/PublicCacheRedisConfiguration.ts` (slight change in the name to prepare the future)

The file `default.config.js` should converted in Typescript to ensure your type definition match Kuzzle default configuration.

Also, you should use the new created type in the `BackendConfig` class

File: lib/config/defaultTsConfig.ts [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+ *
+ * To customize your Kuzzle installation, create a
+ * ".kuzzlerc" file and put your overrides there.
+ * Please check the ".kuzzlerc.sample" file to get
+ * started.
+ *
+ * @class KuzzleConfiguration
+ */
+
+const config: KuzzleConfiguration = {
```
Comment (Aschen) [inline lib/config/defaultTsConfig.ts:17]: Nitpicking: a better name could be `defaultConfig`

File: lib/config/defaultTsConfig.ts [hunk 3]
```diff
@@ -0,0 +1,413 @@
+import httpRoutes from './httpRoutes'
```
Comment (Aschen) [inline lib/config/defaultTsConfig.ts]: Why keeping both the `default.config.js` and this file? You should only have one file for the default configuration and it should be in TS

File: lib/config/index.js [hunk 4]
```diff
@@ -24,7 +24,7 @@
 const assert = require('assert');
 
 const rc = require('rc');
-
+const KuzzleConfiguration = require('../config/defaultTsConfig');
```
Comment (Aschen) [inline lib/config/index.js:27]: You should rather import what you need using the destructuring assignment: `const { defaultConfig } = require(...)`.  
It allows to keep consistent names across the codebase

File: lib/core/backend/backendConfig.ts [hunk 5]
```diff
@@ -25,19 +25,22 @@ import kerror from '../../kerror';
 import kuzzleConfig from '../../config';
 import { JSONObject } from '../../../index';
 import { ApplicationManager, Backend } from './index';
+import { KuzzleConfiguration } from '../../types/config/KuzzleConfiguration';
+import config from '../../config/defaultTsConfig';
+import { load } from '../../config/index.js'
```
Comment (Aschen) [inline lib/core/backend/backendConfig.ts]: This exported function could have a more specific name than `load`

**Review by Aschen — CHANGES_REQUESTED** [review]

File: lib/config/defaultTsConfig.ts [hunk 6]
```diff
@@ -132,15 +145,15 @@ module.exports = {
       profiles: {
         admin: {
           rateLimit: 0,
-          policies: [ { roleId: 'admin'} ]
+          policies: [ { roleID: 'admin'} ]
```
Comment (Aschen) [inline lib/config/defaultTsConfig.ts:148]: This property should stay `roleId` to match Kuzzle internal data representation. This comment apply for others occurrences

File: lib/config/defaultTsConfig.ts [hunk 7]
```diff
@@ -1,4 +1,5 @@
-'use strict';
+import httpRoutes from '../api/httpRoutes.js';
```
Comment (Aschen) [inline lib/config/defaultTsConfig.ts:1]: Why didn't you renamed the file `default.config.ts` ?

File: lib/types/ProfileDefinition.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  /**
+   * Grant a role rights on API actions for this profile.
+   *
+   * Each role can be restricted to a list of indexes and collections.
+   */
+  policies: Array<{
+    /**
+     * Role ID
+     */
+    roleID: string;
```
Comment (Aschen) [inline lib/types/ProfileDefinition.ts:34]: Should be `roleId`

File: lib/types/config/KuzzleConfiguration.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    /**
+     * @default 5000
+     */
+    syncTimeout: number
+  }
+
+  internal: {
+    // hash: {
+    //   seed: number | ArrayBuffer,
+    // }
```
Comment (Aschen) [inline lib/types/config/KuzzleConfiguration.ts:154]: Comment should be removed

**Review by Aschen — CHANGES_REQUESTED** [review]

**Review by Aschen — APPROVED** [review]

**Pull request merged**